### ISSUE TITLE
Cannot use Tree::push_back

### ISSUE THREAD

**Issue by pmolodo** [oooutlk/trees#3]
Trying to use `Tree::push_back`, gives me an error about `DerefMut` not being implemented.

For instance, trying to run this simple code, taken straight from [an example in the docs](https://docs.rs/trees/0.1.0/trees/struct.Node.html#examples-7):

```
fn main() {
    use trees::tr;

    let mut tree = tr(0);
    assert!(tree.is_leaf());
    tree.push_back(tr(1));
    assert!(!tree.is_leaf());
}
```

...gives me this error:

```
error[E0596]: cannot borrow data in a dereference of `trees::linked::fully::tree::Tree<i32>` as mutable
 --> src/main.rs:6:5
  |
6 |     tree.push_back(tr(1));
  |     ^^^^ cannot borrow as mutable
  |
  = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `trees::linked::fully::tree::Tree<i32>`
```

**Comment by oooutlk** [comment]
`<Tree<T> As std::ops::DerefMut>::deref_mut()` will return `&mut Node<T>`, which is unsound, see this [issue](https://github.com/oooutlk/trees/issues/1).

`Tree<T>` will not implement `DerefMut<Target=Node<T>>` since `trees-0.3.0`. And the example code you mentioned is adapted from `trees-0.1.0`.

**Comment by pmolodo** [comment]
I wasn't advocating for implementing `DerefMut` - I was pointing out that:

a) it's not clear if `Tree::push_back` is usable at all in the current version (is it? if so, how?)
b) the docs have an an example which is currently broken

**Comment by oooutlk** [comment]
> a) it's not clear if `Tree::push_back` is usable at all in the current version (is it? if so, how?)

Current version 0.3.0's document does not contain `Tree::push_back()`'s API nor examples. So for users who started with version 0.3.0, they will not be confused about the absence of `Tree::push_back()`. However, for those who started with version 0.1.0, I agree that some explicit declarations would be helpful to make it clear.

By the way, `Tree::push_back()` and other similar APIs could be provided and implemented in delegating to `tree.root_mut()`. I just hesitated whether these API should be provided, and these may be provided in future versions.

> b) the docs have an an example which is currently broken

The example's URL is https://docs.rs/trees/0.1.0/trees/struct.Node.html#examples-7. I tried testing it under [trees-0.1.0](https://crates.io/crates/trees/0.1.0), and it got passed. According to semantic versioning requirements, it is allowed for version 0.3.0 to remove API(s) which were included in version 0.1.0.

Thanks for your mentioning, I'm reconsidering to bring this API back in 0.4.0.

**Comment by pmolodo** [comment]
Hmm, my apologies - for some reason I didn't notice the documentation was for 0.1.0.
I will say, though, that the API is still confusing for people using 0.3.0, since that was the version I was using, and started on.  (I think I originally discovered it via code completion from rust-analyzer.)

**Comment by oooutlk** [comment]
[`Tree::push_back()`](https://github.com/oooutlk/trees/blob/e1e74b4e01a33c35997c856cb1542042b60af489/src/tree.rs#L117) and other APIs are available in [trees 0.4.1](https://crates.io/crates/trees/0.4.1).
